### PR TITLE
[labs/react] Fix for #2799: add support for properties with custom accessors

### PR DESCRIPTION
**Pull request by lideen**
Hi! All new to this repo but tried to fix a bug that I found #2799

Any input appreciated 🎉

### CODE REVIEW

**Comment by justinfagnani** [comment]
@lideen thanks for identifying this and submitting a fix and tests!

I made some comments on the implementation in the interest of speed here. I think the implementation can be collapsed into one `getInstancePropertiesOf()` function that recursively calls itself on a class's superclass, merges the superclass's properties into its own, then caches and returns the result. This will avoid materializing the prototype chain and processing prototypes multiple times. It should minimize memory allocations too.

I would like to also avoid materializing the property descriptors if possible. I think the core issue here is that accessors are not enumerable and `for...in` doesn't include non-enumerable properties. `Object.getOwnPropertyNames()` does return non-enumerable names (`Reflect.ownKeys()`) does too, but it's not supported on IE11 and we don't need Symbol-keyed properties in JSX). So I hope we can use that. It would include methods, but that might be ok.

**Comment by lideen** [comment]
@justinfagnani Something like this? 1531750aedc7b4c8cbe4a39ad414fe301a496cc9

**Comment by justinfagnani** [comment]
@lideen that what I was thinking at the time (though with memoizing the function).

But in the meantime I chatted with @josepharhar about it, and I think that we should match the Preact and upcoming React support for properties as much as possible. This would mean not pre-calculating the instance keys at all, but just using an `in` check on each given prop key. See https://github.com/facebook/react/blob/2c8a1452b82b9ec5ebfa3f370b31fda19610ae92/packages/react-dom/src/client/DOMPropertyOperations.js#L187-L194

**Comment by lideen** [comment]
@justinfagnani I gave it a shot [79e6a4f](https://github.com/lit/lit/commit/79e6a4f793bd9cbb1dea58ca387498abe7f56844). I removed the warning about reserved props as I don't think it fits well now when the check is done on-demand 🤔 . Let me know if I should re-add it.

**Comment by lideen** [comment]
@justinfagnani sorry to ping you, but is there anything I can do to move this along?

File: packages/labs/react/src/create-component.ts [hunk 1]
```diff
@@ -257,7 +236,12 @@ export const createComponent = <I extends HTMLElement, E extends Events>(
       for (const [k, v] of Object.entries(this.props)) {
         if (k === '__forwardedRef') continue;
 
-        if (elementClassProps.has(k)) {
+        if (
+          eventProps.has(k) ||
+          (!reservedReactProperties.has(k) &&
+            !(k in HTMLElement.prototype) &&
+            k in elementClass.prototype)
```
Comment (justinfagnani) [inline packages/labs/react/src/create-component.ts:243]: We still need to verify this logic (specifically excluding `k in HTMLElement.prototype`) with @josepharhar to see if it matches React 19. I'm good with going in as-is now though.

File: packages/labs/react/src/create-component.ts [hunk 2]
```diff
@@ -157,28 +158,6 @@ export const createComponent = <I extends HTMLElement, E extends Events>(
...
-  const elementClassProps = new Set(Object.keys(events ?? {}));
-  for (const p in elementClass.prototype) {
-    if (!(p in HTMLElement.prototype)) {
-      if (reservedReactProperties.has(p)) {
-        // Note, this effectively warns only for `ref` since the other
-        // reserved props are on HTMLElement.prototype. To address this
-        // would require crawling down the prototype, which doesn't feel worth
-        // it since implementing these properties on an element is extremely
-        // rare.
-        console.warn(
```
Comment (justinfagnani) [inline packages/labs/react/src/create-component.ts:171]: I think we should keep this warning. We can move this into a dev build in the future.

**Review by justinfagnani — CHANGES_REQUESTED** [review]

**Comment by justinfagnani** [comment]
@lideen this has increased in urgency for us, so I'm going to try to see if I can get this in asap. Did you enable changes to your PR branch? https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

**Review by justinfagnani — APPROVED** [review]
Thanks @lideen !

I'll follow up on the comments I left with issues.

**Pull request merged**